File: pr6_project/parser.c
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <stdlib.h>
#include "formatter.h"
#include "parser.h"

/**
 * Checks if the string is a valid message for OLDM
 * [OLDM␣num-mess␣id␣message] 161 bytes
 */
int checkOLDM(char *str)
{
    // Check spaces
    if (str == NULL || strlen(str) != 161 || str[9] != ' ' || str[18] != ' ')
        return 0;
    // Check keywords
    if (memcmp(str, "OLDM ", 5) || memcmp(str + 159, "\r\n", 2))
        return 0;
    char num[5];
    num[4] = '\0';
    memcpy(num, str + 5, 4);
    char *num_test = unFormatNumber(num);
    if (num_test == NULL)
        return 0;
    free(num_test);
    return 1;
}

/**
 * Checks if the string is a valid message for LINB
 * [LINB␣num-diff] 9 bytes
 */
int checkLINB(char *str)
{
    // Check spaces
    if (str == NULL || strlen(str) != 9)
        return 0;
    // Check keywords
    if (memcmp(str, "LINB ", 5) || memcmp(str + 7, "\r\n", 2))
        return 0;
    char num[3];
    num[2] = '\0';
    memcpy(num, str + 5, 2);
    char *num_test = unFormatNumber(num);
    if (num_test == NULL)
        return 0;
    free(num_test);
    return 1;
}

/**
 * Checks if the string is a valid message for ITEM
 * [ITEM␣id␣ip1␣port1␣ip2␣port2] 57 bytes
 */
int checkITEM(char *str)
{
    // Check spaces
    if (str == NULL || strlen(str) != 57 || str[13] != ' ' || str[29] != ' ' ||
        str[34] != ' ' || str[50] != ' ')
        return 0;
    // Check keywords
    if (memcmp(str, "ITEM ", 5) || memcmp(str + 55, "\r\n", 2))
        return 0;
    // Get IPs and Ports
    char ip1[16];
    ip1[15] = '\0';
    memcpy(ip1, str + 14, 15);
    char port1[5];
    port1[4] = '\0';
    memcpy(port1, str + 30, 4);
    char ip2[16];
    ip2[15] = '\0';
    memcpy(ip2, str + 35, 15);
    char port2[5];
    port2[4] = '\0';
    memcpy(port2, str + 51, 4);

    // Check IPs
    if (!isValidIPv4Format(ip1) || !isValidIPv4Format(ip2))
        return 0;
    // Check Ports
    char *port1_test = unFormatNumber(port1);
    if (port1_test == NULL)
        return 0;
    free(port1_test);
    char *port2_test = unFormatNumber(port2);
    if (port2_test == NULL)
        return 0;
    free(port2_test);

    return 1;
}

/**
 * Checks if the string is a valid message for DIFF
 * [DIFF␣num-mess␣id␣message] 161 bytes
 */
int checkDIFF(char *str)
{
    // Check spaces
    if (str == NULL || strlen(str) != 161 || str[9] != ' ' || str[18] != ' ')
        return 0;
    // Check keywords
    if (memcmp(str, "DIFF ", 5) || memcmp(str + 159, "\r\n", 2))
        return 0;
    char num[5];
    num[4] = '\0';
    memcpy(num, str + 5, 4);
    char *num_test = unFormatNumber(num);
    if (num_test == NULL)
        return 0;
    free(num_test);
    return 1;
}

/**
 * Test purposes (Simple tests functions above)
 * Uncomment the main to test
 * gcc -g3 -Wall parser.c formatter.c formatter.h
 * ./a.out
 */
/*
int main(int argc, char **argv)
{
    // Init Messages Values
    char mess[141];
    mess[140] = '\0';
    memset(mess, 'a', 140);

    char oldmOK[162] = "OLDM 0012 RADIO### ";
    strcat(oldmOK, mess);
    strcat(oldmOK, "\r\n\0");
    char linbOK[10] = "LINB 02\r\n\0";
    char itemOK[58] = "ITEM RADIO### 127.000.000.001 5551 225.001.002.004 5552\r\n\0";
    char diffOK[162] = "DIFF 0012 RADIO### ";
    strcat(diffOK, mess);
    strcat(diffOK, "\r\n\0");

    printf("Test OLDM OK %d\n", checkOLDM(oldmOK));
    printf("Test LINB OK %d\n", checkLINB(linbOK));
    printf("Test ITEM OK %d\n", checkITEM(itemOK));
    printf("Test DIFF OK %d\n", checkDIFF(diffOK));

    char *oldmNO = "OLDM OLDN 0012 RADIO### aaaa\r\n\0"; // Missing big part of mess
    char *linbNO = "LINB 02 \n\0"; // Missing \r
    char *itemNO = "ITEM RRADIO### 127.000.000.001 5551 225.001.002.004 5552\r\n\0"; // Extra char
    char diffNO[162] = "DIFF 0012 RADIO### ";
    strcat(diffNO, mess);
    strcat(diffNO, "\r\0"); // Missing \n

    printf("\nTest OLDM NO %d\n", checkOLDM(oldmNO));
    printf("Test LINB NO %d\n", checkLINB(linbNO));
    printf("Test ITEM NO %d\n", checkITEM(itemNO));
    printf("Test DIFF NO %d\n", checkDIFF(diffNO));
}
 */

File: pr6_project/Parser.java
import java.io.*;
public class Parser {

    /**
     * Private Constructor
     */
    private Parser() {}

    /**
     * This function reads from the buffered reader and
     * returns a String ending with "\r\n"
     *
     * @param br BufferedReader
     * @return String
     */
    public static String readFullLine(BufferedReader br) {
        String mess = "";
        char cr = '\0'; // Init Carriage Return
        char c; // Init c
        try {
            // A Message MUST END with "\r\n" as said in subject
            while (((c = (char) br.read()) != '\n') || (cr != '\r')) {
                if (!br.ready() && c != '\n' && cr != '\r')
                    throw new IOException("Message did not end with \"\\r\\n\"");
                cr = c;
                mess += c;
            }
            mess += c;
        } catch (IOException e) {
            System.out.println(e.getMessage());
            return "";
        }
        return mess;
    }

    /**
     * Checks if the string is a valid message for DIFF
     * [DIFF␣num-mess␣id␣message] 161 bytes
     * @param str
     * @return boolean
     */
    public static boolean checkDIFF(String str) {
        // Check spaces
        if (str.length() != 161 || str.charAt(9) != ' ' || str.charAt(18) != ' ')
            return false;
        // Check keywords
        if (!str.startsWith("DIFF ") || !str.endsWith("\r\n"))
            return false;
        try {
            Formatter.unFormatNumber(str.substring(5, 9)); // Check num-mess
        } catch (Exception e) {
            return false;
        }
        return true;
    }

    /**
     * Checks if the string is a valid message for REGI
     * [REGI␣id␣ip1␣port1␣ip2␣port2] 57 bytes
     *
     * @param str
     * @return boolean
     */
    public static boolean checkREGI(String str) {
        // Check spaces
        if (str.length() != 57 || str.charAt(13) != ' ' || str.charAt(29) != ' ' ||
            str.charAt(34) != ' ' || str.charAt(50) != ' ')
            return false;
        // Check keywords
        if (!str.startsWith("REGI ") || !str.endsWith("\r\n"))
            return false;
        String tab[] = str.substring(5, 55).split(" ");
        if (tab.length != 5 || tab[0].length() != 8 || tab[1].length() != 15 ||
            tab[2].length() != 4 || tab[3].length() != 15 || tab[4].length() != 4)
            return false;
        try {
            // Check IPs
            if (!Formatter.formatIPv4(tab[1]).equals(tab[1]) ||
                !Formatter.formatIPv4(tab[3]).equals(tab[3])) {
                    System.out.println("FAIL TRY");
                    return false;
                }
            // Check Ports
            Formatter.unFormatNumber(tab[2]);
            Formatter.unFormatNumber(tab[4]);
        } catch (Exception e) {
            return false;
        }
        return true;
    }

    /**
     * Checks if the string is a valid message for MESS
     * [MESS␣id␣message] 156 bytes
     *
     * @param str
     * @return boolean
     */
    public static boolean checkMESS(String str) {
        // Check spaces
        if (str.length() != 156 || str.charAt(13) != ' ')
            return false;
        // Check keywords
        if (!str.startsWith("MESS ") || !str.endsWith("\r\n"))
            return false;
        return true;
    }

    /**
     * Checks if the string is a valid message for LAST
     * [LAST␣nb-mess] 10 bytes
     *
     * @param str
     * @return boolean
     */
    public static boolean checkLAST(String str) {
        // Check keywords
        if (str.length() != 10 || !str.startsWith("LAST ") || !str.endsWith("\r\n"))
            return false;
        try {
            Formatter.unFormatNumber(str.substring(5, 8));
        } catch (Exception e) {
            return false;
        }
        return true;
    }

    // Test purposes (Simple tests functions above)
    public static void main(String[] args) {
        // Init Messages Values
        String mess = "";
        for (int i = 0; i < 140; i++)
            mess += 'a';
        String id = "RADIO###";
        String ip1 = "127.000.000.001";
        String ip2 = "225.001.002.004";
        String num_mess = "0012";
        String nb_mess = "098";
        String port1 = "5551";
        String port2 = "5552";

        String diffOK = "DIFF " + num_mess + " " + id + " " + mess + "\r\n";
        String regiOK = "REGI " + id + " " + ip1 + " " + port1 + " " + ip2 + " " + port2 + "\r\n";
        String messOK = "MESS " + id + " " + mess + "\r\n";
        String messOK2 =
            "MESS channele Sed ut perspiciatis unde omnis iste natus error sit " +
            "voluptatem accusantium dol1########################################" +
            "#####################\r\n";
        String lastOK = "LAST " + nb_mess + "\r\n";

        System.out.println("Test DIFF OK " + checkDIFF(diffOK));
        System.out.println("Test REGI OK " + checkREGI(regiOK));
        System.out.println("Test MESS OK " + checkMESS(messOK));
        System.out.println("Test MESS OK2 " + checkMESS(messOK2));
        System.out.println("Test LAST OK " + checkLAST(lastOK));

        String diffNO = "DIFF " +  "0a23 " + id + " " + mess + "\r\n"; // num_mess not number
        String regiNO = "REGI" + id + " " + ip1 + " " + port1 + " " + ip2 + " " + port2 + "\n"; // Missing end
        String messNO = "MESS " + id + " " + mess + "\r\n "; // Extra Space at the end
        String lastNO = "LAST " + nb_mess + "  \n"; // Space instead of carriage return

        System.out.println("\nTest DIFF NO " + checkDIFF(diffNO));
        System.out.println("Test REGI NO " + checkREGI(regiNO));
        System.out.println("Test MESS NO " + checkMESS(messNO));
        System.out.println("Test LAST NO " + checkLAST(lastNO));
    }
}